libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -188,43 +581,200 @@ proc getGossipPeers(g: GossipSub): Table[PubSubPeer, ControlMessage] {.gcsafe.}
 
   trace "getting gossip peers (iHave)"
   let topics = toHashSet(toSeq(g.mesh.keys)) + toHashSet(toSeq(g.fanout.keys))
-  let controlMsg = ControlMessage()
   for topic in topics:
-    var allPeers = toSeq(g.gossipsub.getOrDefault(topic))
-    shuffle(allPeers)
-
-    let mesh = g.mesh.getOrDefault(topic)
-    let fanout = g.fanout.getOrDefault(topic)
+    if topic notin g.gossipsub:
+      trace "topic not in gossip array, skipping", topicID = topic
+      continue
 
-    let gossipPeers = mesh + fanout
     let mids = g.mcache.window(topic)
     if not mids.len > 0:
       continue
 
-    if topic notin g.gossipsub:
-      trace "topic not in gossip array, skipping", topic
-      continue
-
     let ihave = ControlIHave(topicID: topic, messageIDs: toSeq(mids))
+
+    let mesh = g.mesh.getOrDefault(topic)
+    let fanout = g.fanout.getOrDefault(topic)
+    let gossipPeers = mesh + fanout
+    var allPeers = toSeq(g.gossipsub.getOrDefault(topic))
+
+    allPeers.keepIf do (x: PubSubPeer) -> bool:
+      x.peerId notin g.parameters.directPeers and
+      x notin gossipPeers and
+      x.score >= g.parameters.gossipThreshold
+
+    var target = g.parameters.dLazy
+    let factor = (g.parameters.gossipFactor.float * allPeers.len.float).int
+    if factor > target:
+      target = min(factor, allPeers.len)
+
+    if target < allPeers.len:
+      shuffle(allPeers)
+      allPeers.setLen(target)
+
     for peer in allPeers:
-      if result.len >= GossipSubD:
-        trace "got gossip peers", peers = result.len
-        break
+      if peer notin result:
+        result[peer] = ControlMessage()
+      result[peer].ihave.add(ihave)
+
+func `/`(a, b: Duration): float64 =
+  let
+    fa = float64(a.nanoseconds) / 1000000000
```

Comment:
why the division by 10000000?

Reply:
Something related to this 
https://github.com/status-im/nim-chronos/blob/21349807444b86f9ce7e2b533a3ecc690f3edc3d/chronos/timer.nim#L185

Reply:
1 sec = 1e+9 and all that jazz

Reply:
eeh... basic math though?

Reply:
lol yeah it's pointless scaling, I have no idea why I did that 😮

Reply:
fixed

Reply:
I swear there must have been a reason at some point 🔢 😄